Playwright: migrate WordPress.com Paid Signup tests.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates the standard Paid portion of the Signup spec.

Key changes:
- implemented methods and data to test the store.
- new methods in the `CartCheckoutPage` to interact with the billing and payment sections.
- new flow to cancel purchases.
- refactored the account closure steps to a flow.
- updated the store testing cookie value in configuration file.

Background details:

This PR builds on top of #56087 and combines several paths of the existing Selenium spec to test the following features:
- purchase of a premium plan during the signup process.
- cancellation of a plan.
- launching the site.

The overall look and feel will feel very similar to the Signup: Free spec.

#### Testing instructions

regular e2e tests
   - [x] mobile
   - [x] desktop


Related to #

worldomonation committed:
Implement skeleton of Signup: Paid

worldomonation committed:
Finish migration

Fix linting error.

worldomonation commented:
This PR is ready to review.

Note, this test is not yet run by default (tagged for pre-release only).

worldomonation commented on packages/calypso-e2e/src/browser-manager.ts:
```diff
@@ -116,3 +116,42 @@ export async function clearAuthenticationState( page: Page ): Promise< void > {
+ * @param {Page} page Object representing a page launched by Playwright.
+ * @param {string} currency ISO 4217-compliant currency code.
+ */
+export async function setStoreCookie(
```
Instead of having multiple methods to set different cookies, enable additional cookies using flags.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -141,6 +155,31 @@ export function getTestEmailAddress( {
+ * @returns {[key: string]: string} Keyed object containing strings as values.
+ */
+export function getTestPaymentDetails(
+	issuer: 'Visa' | 'Mastercard' | 'American Express'
```
Stub for future date for other simulated payment methods.

worldomonation commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -2,7 +2,7 @@ import { Page } from 'playwright';
 const selectors = {
 	// Buttons on navbar
-	mySiteButton: 'button[data-tip-target="my-sites"]',
+	mySiteButton: '[data-tip-target="my-sites"]',
```
On most of Calypso, this is indeed a button.

However, after launching the site, this becomes a hyperlink instead.

worldomonation commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -1,11 +1,42 @@
+	// Payment field
+	cardholderName: `input[id="cardholder-name"]`,
+	cardNumberFrame: 'iframe[title="Secure card number input frame"]',
+	cardNumberInput: 'input[data-elements-stable-field-name="cardNumber"]',
```
Within the card number iframe there are multiple hidden input tags, so we must explicitly look for the correct field.

worldomonation commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -42,4 +73,105 @@ export class CartCheckoutPage {
+		await this.page.click( selectors.couponCodeInputButton );
+
+		const couponCode =
+			coupon !== undefined ? coupon : ( DataHelper.config.get( 'testCouponCode' ) as string );
```
My thought here against what I've done, is that this method should not be responsible for looking up a coupon code if not passed in.

dpasque replied:
Yeah, I agree, that feels like unneeded complexity to me! I think it would be better to expose a method in `DataHelper` called `getTestCouponCode`, then users can call that and then feed that to this function. That modularity might make both of those functions more re-usable too!

worldomonation replied:
Alternatively, what if we expect the test to feed in the coupon code?

worldomonation commented on packages/calypso-e2e/src/lib/pages/my-home-page.ts:
```diff
@@ -2,6 +2,9 @@ import { Page } from 'playwright';
 	visitSiteButton: '.button >> text=Visit site',
+
+	// Task card (topmost card)
+	taskMessage: ( message: string ) => `div.task h2:has-text("${ message }")`,
```
Generic selector that will match the heading item shown in My Home dashboard.

dpasque replied:
In the function comments, you say it checks for a string on the card, but this selector limits it to just the header. I think we should clarify either way and lock that in more cleary!

worldomonation replied:
I thought I had specified it was only the header. I'll review and update the comments as necessary.

worldomonation commented on test/e2e/specs/specs-playwright/wp-signup__free.ts:
```diff
@@ -117,21 +114,9 @@ describe(
-			it( 'Confirm account is closed', async function () {
-				const accountClosedPage = new AccountClosedPage( page );
-				await accountClosedPage.confirmAccountClosed();
+			it( 'Close account', async function () {
+				const closeAccountFlow = new CloseAccountFlow( page );
+				await closeAccountFlow.closeAccount();
```
Share the same flow which has been extracted into its own Flow object.

worldomonation commented on test/e2e/specs/specs-playwright/wp-signup__paid.ts:
```diff
@@ -0,0 +1,183 @@
+			const testContext = await BrowserManager.browser.newContext();
+			const testPage = await testContext.newPage();
+			// TODO: make a utility to obtain the blog URL.
+			await testPage.goto( `https://${ blogName }.wordpress.com` );
```
In this step, instead of clearing the authenticated state for this Page instance I have chosen to launch an entirely new context (private session window). This allows the test spec to check whether the site is hidden behind a `Coming Soon` banner without having to re-log in.

worldomonation replied:
In the future, it may be a good idea to have an accessible method within BrowserManager to launch a new context. There exists `launchContext` but I recall that it I could not get it to work in this spec.

dpasque replied:
Do you want to make a task to come back around to do this? I think that would be a worthwhile modularization to make in the future! Looking at the `BrowserManager`, it looks like we _could_ use `launchPage`, but we'd have to do some refactoring around the `Logger`.

worldomonation replied:
Sure, let's do that. The `Logger` parameter is really throwing me off, and there were other instances where I could've used the `launchXYZ` methods but was thwarted.

worldomonation replied:
Created #56363 for this.

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -141,6 +155,31 @@ export function getTestEmailAddress( {
+ * @returns {[key: string]: string} Keyed object containing strings as values.
+ */
+export function getTestPaymentDetails(
+	issuer: 'Visa' | 'Mastercard' | 'American Express'
```
For now, for safety, want to just type this as `'Visa'`? It's really easy for folks to just extend that type when we want to add more cards. That seems better than someone getting an unexpected null, and you could take the null out of the function signature!

worldomonation replied:
You're right. I'll remove other card types for now.

worldomonation replied:
Fixed in e16661468d2dcd16127a4850c2cf7e0fc289e78c.

dpasque commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -42,4 +73,105 @@ export class CartCheckoutPage {
+		if ( ! parsedAmount?.length ) {
+			throw new Error( 'Unable to locate or parse cart amount.' );
+		}
+		return parseInt( parsedAmount.pop() as string );
```
Out of curiosity, why `parseInt` instead of `parseFloat`? With the coupon, you can definitely get cents in American currency at least! And we're matching the decimals in the RegEx!

worldomonation replied:
I carried this over from Selenium, where they discarded all decimal points in the comparison.

The idea is that since the test coupon code provides 10% off, the price (36GBP) would change the digit in the ones-place with the coupon code applied.

dpasque commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -42,4 +73,105 @@ export class CartCheckoutPage {
+	/**
+	 * Returns the total amount of items in cart.
+	 *
+	 * @returns {string} Total value of items in cart.
```
Do we need to make this i18n safe? In the cookies function, you allow for other currencies, but this current function wouldn't be able to some i18n currency options (mostly due to separator differences).

I think that's fine if we don't want to for now, but we should make that very clear in the docs or function name that it may not work for all currency separators!

worldomonation replied:
My guess is you're talking about continental Europe where they use a full stop for separating thousands and a comma for the decimal place?

You are correct, that would mess this up. I don't know of a good way to handle that - honestly it's a mess (digit grouping) and I wish we just didn't do it.

worldomonation replied:
My proposal is to leave this to a future time when that becomes necessary.

In our testing so far, we haven't used anything other than GBP and AUD. There is a point to be made about testing other currencies, but let's cross that bridge when we need to.

dpasque commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -42,4 +73,105 @@ export class CartCheckoutPage {
+	 * @param {string} param0.expiryYear Credit card expiry year.
+	 * @param {string} param0.cvv Credit card CVV value.
+	 */
+	async enterPaymentDetails( {
+		cardHolder,
+		cardNumber,
+		expiryMonth,
+		expiryYear,
+		cvv,
+	}: PaymentDetails ): Promise< void > {
```
Nit incoming! Since you have an interface for this, I think it would be more readable to just take a single named parameter of the `PaymentDetails` type, and then you can destructure it in the function. That should simplify the signature / header comments a good bit.

worldomonation replied:
Fixed in e16661468d2dcd16127a4850c2cf7e0fc289e78c.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__paid.ts:
```diff
@@ -0,0 +1,181 @@
+			// eg. USD would commonly have 2 decimal places, e.g. 12.34.
+			// In JPY or TWD there will be no decimal digits.
+			// Drop decimals so that the result won't be affected by the currency variation.
+			expect( Math.floor( newAmount ) ).toStrictEqual( Math.floor( expectedAmount ) );
```
I actually feel like while it's good to handle this variability in the library, we don't need to do handle this here, and can just tailor this test case to whatever we expect for GBP. The test spec itself inevitably has to be coupled to a given currency, and I think that's okay!

worldomonation replied:
Could you clarify what you mean?

Instead of performing the math operation here, move it into the library?

dpasque reviewed (commented):
Left a few smaller comments, but overall this is looking great!

The test flow here is so clean and easy to follow, and covers so much critical ground in a few steps. I love it! 😄

worldomonation committed:
Address PR comments:
- refactor `getTestPaymentDetails` to eliminate non-Visa options for now.
- change `enterCouponCode` method to expect a coupon code to be supplied.
- clarify taskMessage.
- simplify function signature on payment details

dpasque reviewed (approved):
The currency handling should work for now, we can always revisit later too, so approving!

worldomonation committed:
Change returned value of `getCheckoutTotalAmount` call.
Note in the comment for the method, that it only works for locales that use a full stop for decimal.

worldomonation committed:
Post-checkout quickstart upsell may not occur all the time, so remove the steps.
Adjust the plan cancellation screen due to #56206.

worldomonation committed:
Rebase and incorporate fixes from #56321.
Add waits until coupon application is complete and make banner dismissal an optional step.

worldomonation committed:
Use different names for paid and free signup tests to eliminate possibility of clashes.
Search for field for secure credit card input.

worldomonation commented:
Yay, the Pre-Release tests are now passing!
